src/hotspot/os/linux/memMapPrinter_linux.cpp

```diff
@@ -25,60 +25,165 @@
+#include "utilities/powerOfTwo.hpp"
+
 #include <limits.h>
 
-struct ProcMapsInfo {
-  void* from = 0;
-  void* to = 0;
-  char prot[20 + 1];
-  char offset[20 + 1];
-  char dev[20 + 1];
-  char inode[20 + 1];
-  char filename[1024 + 1];
-
-  bool scan_proc_maps_line(const char* line) {
-    prot[0] = offset[0] = dev[0] = inode[0] = filename[0] = '\0';
-    const int items_read = ::sscanf(line, "%p-%p %20s %20s %20s %20s %1024s",
-        &from, &to, prot, offset, dev, inode, filename);
-    return items_read >= 2; // need at least from and to
+class ProcSmapsSummary {
+  unsigned _num_mappings;
+  size_t _vsize;        // combined virtual size
+  size_t _rss;          // combined resident set size
+  size_t _committed;    // combined committed space
+  size_t _shared;       // combined shared size
+  size_t _swapped_out;  // combined amount of swapped-out memory
+  size_t _hugetlb;      // combined amount of memory backed by explicit huge pages
+  size_t _thp;          // combined amount of memory backed by THPs
+public:
+  ProcSmapsSummary() : _num_mappings(0), _vsize(0), _rss(0), _committed(0), _shared(0),
+                     _swapped_out(0), _hugetlb(0), _thp(0) {}
+  void add_mapping(const ProcSmapsInfo& info) {
+    _num_mappings++;
+    _vsize += info.vsize();
+    _rss += info.rss;
+    _committed += info.nr ? 0 : info.vsize();
+    _shared += info.sh ? info.vsize() : 0;
+    _swapped_out += info.swap;
+    _hugetlb += info.private_hugetlb + info.shared_hugetlb;
+    _thp += info.anonhugepages;
+  }
+
+  void print_on(const MappingPrintSession& session) const {
+    outputStream* st = session.out();
+    st->print_cr("Number of mappings: %u", _num_mappings);
+    st->print_cr("             vsize: %zu (" PROPERFMT ")", _vsize, PROPERFMTARGS(_vsize));
+    st->print_cr("               rss: %zu (" PROPERFMT ")", _rss, PROPERFMTARGS(_rss));
+    st->print_cr("         committed: %zu (" PROPERFMT ")", _committed, PROPERFMTARGS(_committed));
+    st->print_cr("            shared: %zu (" PROPERFMT ")", _shared, PROPERFMTARGS(_shared));
+    st->print_cr("       swapped out: %zu (" PROPERFMT ")", _swapped_out, PROPERFMTARGS(_swapped_out));
+    st->print_cr("         using thp: %zu (" PROPERFMT ")", _thp, PROPERFMTARGS(_thp));
+    st->print_cr("           hugetlb: %zu (" PROPERFMT ")", _hugetlb, PROPERFMTARGS(_hugetlb));
   }
 };
 
-class LinuxMappingPrintInformation : public MappingPrintInformation {
-  const ProcMapsInfo _info;
+class ProcSmapsPrinter {
+  const MappingPrintSession& _session;
 public:
+  ProcSmapsPrinter(const MappingPrintSession& session) :
+    _session(session)
+  {}
 
-  LinuxMappingPrintInformation(const void* from, const void* to, const ProcMapsInfo* info) :
-    MappingPrintInformation(from, to), _info(*info) {}
+  void print_single_mapping(const ProcSmapsInfo& info) const {
+    outputStream* st = _session.out();
+#define INDENT_BY(n)          \
+  if (st->fill_to(n) == 0) {  \
```

Comment:
`fill_to` returns `void`, am I missing something?

Reply:
It was changed in this PR to return the number of spaces actually printed.